concurrency/ServiceController.java

```diff
@@ -0,0 +1,1922 @@
+                AccountSettingsUtils.increaseSyncWindow(context, mailbox);
+                ServiceController.startSync(context, mailboxId, ServiceController.SYNC_UI_REFRESH);
+            }
+        });
+    }
+
+    private static final String UNSET_WARN_FLAG_SELECT = Message.MAILBOX_KEY + "=?";
+
+    /**
+     * Unset attachment warning flags for all messages in outbox
+     */
+    public static void unsetAttachmentSizeWarnings(final Context context, final long accountId) {
+        // FIXME: This call was runAsyncParallel, to reduce rejectedExecutionException
+        // the call is made serial. Refer: PIM-17503
+        EmailAsyncTask.runAsyncSerial(new Runnable() {
+            @Override
+            public void run() {
+                final long outboxId = Mailbox.findMailboxOfType(context, accountId,
+                        Mailbox.TYPE_OUTBOX);
+                if (outboxId == Mailbox.NO_MAILBOX) {
+                    return;
+                }
+                ContentResolver resolver = context.getContentResolver();
+                resolver.update(EmailContent.Message.UNSET_WARN_FLAG_URI, null,
+                        UNSET_WARN_FLAG_SELECT, new String[] { Long.toString(outboxId) });
+            }
+        });
+    }
+
+    public static boolean isPushInProgress(Context context, long mailboxId) {
+        ServiceController controller = sInstance;
+        if (controller == null) {
+            // By definition, it's not... :-)
+            return false;
+        }
+        ConcurrentHashMap<Long, Long> map = controller.mManagedMailboxMap;
+        if (map.containsKey(mailboxId)) {
+            mailboxId = map.get(mailboxId);
+        }
+        // Ok, this is the mailboxId of the controlling mailbox
+        ServiceEntry entry = controller.mServiceMap.get(mailboxId);
+        return (entry != null) && (entry.mError == ServiceEntry.NO_ERROR) &&
+            (entry.mSyncInterval == Mailbox.CHECK_INTERVAL_PUSH);
+    }
+
+    public static String getSyncReason(int reason) {
+        if (reason >= 0 && reason < SYNC_REASONS.length) {
+            return SYNC_REASONS[reason];
+        }
+        return "None";
+    }
+
+    public static void breakPing(Context context, long accountId) {
+        Mailbox m = Mailbox.getMailboxForType(context, accountId, Mailbox.TYPE_INBOX);
+        startSync(context, m.mId, SYNC_REASON_ANY);
+    }
+
+    private EmailServiceProxy getLiveProxy(String protocol) {
+        synchronized(sConnectivityLock) {
+            EmailServiceProxy proxy = mProxyCache.get(protocol);
+            // treat dead proxies in our cache as if they weren't there
+            return (proxy != null && proxy.isDead()) ? null : proxy;
+        }
+    }
+
+    private EmailServiceProxy removeProxyIfDead(String protocol) {
+        synchronized(sConnectivityLock) {
+            EmailServiceProxy proxy = mProxyCache.get(protocol);
+            return (proxy != null && proxy.isDead()) ? mProxyCache.remove(protocol) : null;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **1857 and 1858**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `remove(key)` with `remove(key, value)` and then removing the synchronization. The check is at line: **1858**. If you do not need the check, consider removing the check and the synchronization and just calling `remove(key)`.

**More info**  
 `remove(key, value)` removes the key only if the current value is equal to the `value` parameter. `remove(key, value)` returns `true` if the call removed the `key` and `value`.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/4d3c580b03475a6ec9323d11e6875c542f8e3f6d/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java#L1591-L1592) (external link).